taroscript/send.go

```diff
@@ -446,202 +347,301 @@ func CompleteAssetSpend(internalKey btcec.PublicKey, delta SpendDelta,
+	// Sanity check the output indexes provided by the sender. There must be
+	// at least two indexes; one for the receiver, and one for the change
+	// commitment for the sender.
+	if len(outputs) < 2 {
+		return false, ErrInvalidOutputIndexes
+	}
 
-		receiverCommitment, err = commitment.NewAssetCommitment(
-			receiverAssetCopy,
-		)
-		if err != nil {
-			return nil, err
+	// If the indexes start from 0 and form a continuous range, then the
+	// resulting TX would be valid without any changes (Taro-only spend).
+	taroOnlySpend := true
+	sortedCopy := slices.Clone(outputs)
+	sort.Slice(sortedCopy, func(i, j int) bool {
+		return sortedCopy[i].AnchorOutputIndex <
+			sortedCopy[j].AnchorOutputIndex
+	})
+	for i := 0; i < len(sortedCopy); i++ {
+		if sortedCopy[i].AnchorOutputIndex != uint32(i) {
+			taroOnlySpend = false
+			break
 		}
 	}
 
-	// Update the top-level TaroCommitment of the sender. This'll
-	// effectively commit to all the new spend details.
-	//
-	// TODO(jhb): Add emptiness check for senderCommitment, to prune the
-	// AssetCommitment entirely when possible.
-	senderTaroCommitment := *inputCommitmentCopy
-	err = senderTaroCommitment.Update(senderCommitment, false)
+	return taroOnlySpend, nil
+}
+
+// CreateAnchorTx creates a template BTC anchor TX with dummy outputs.
+func CreateAnchorTx(outputs []*taropsbt.VOutput) (*psbt.Packet, error) {
+	// Check if our outputs are valid, and if we will need to add extra
+	// outputs to fill in the gaps between outputs.
+	taroOnlySpend, err := AreValidAnchorOutputIndexes(outputs)
 	if err != nil {
 		return nil, err
 	}
 
-	commitments[senderStateKey] = senderTaroCommitment
+	// Calculate the number of outputs we need for our template TX.
+	maxOutputIndex := uint32(len(outputs))
 
-	// Create a Taro tree for the receiver.
-	receiverTaroCommitment, err := commitment.NewTaroCommitment(
-		receiverCommitment,
-	)
+	// If there is a gap in our outputs, we need to find the
+	// largest output index to properly size our template TX.
+	if !taroOnlySpend {
+		maxOutputIndex = 0
+		for _, out := range outputs {
+			if out.AnchorOutputIndex > maxOutputIndex {
+				maxOutputIndex = out.AnchorOutputIndex
+			}
+		}
+
+		// Output indexes are 0-indexed, so we need to increment this
+		// to account for the 0th output.
+		maxOutputIndex++
+	}
+
+	txTemplate := wire.NewMsgTx(2)
+	for i := uint32(0); i < maxOutputIndex; i++ {
+		txTemplate.AddTxOut(createDummyOutput())
+	}
+
+	spendPkt, err := psbt.NewFromUnsignedTx(txTemplate)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("unable to make psbt packet: %w", err)
 	}
 
-	commitments[receiverStateKey] = *receiverTaroCommitment
+	for i := range outputs {
+		out := outputs[i]
+		psbtOut := &spendPkt.Outputs[out.AnchorOutputIndex]
+		psbtOut.TaprootInternalKey = schnorr.SerializePubKey(
+			out.AnchorOutputInternalKey,
+		)
+		if out.AnchorOutputBip32Derivation != nil {
```

Comment:
IIUC these can be nil from non-local keys given in an address, is there another case where that would be true?

Reply:
Yes, for non-local keys, mostly. Or for malformed/incomplete packets, since this could come from an RPC. So we'd rather be cautious and fail with a "missing derivation information" error later on.